### PR TITLE
Add Support for Sending Events

### PR DESCRIPTION
**Pull request by xshipeng**

### CODE REVIEW

File: src/main/java/com/wavefront/sdk/common/Utils.java [hunk 1]
```diff
@@ -260,6 +263,132 @@ public static String tracingSpanToLineData(String name, long startMillis, long d
...
+    // check and sanitize parameters
+    if (source == null || source.isEmpty()) {
+      source = defaultSource;
+    }
+    if (name == null || name.isEmpty()) {
+      throw new IllegalArgumentException("event name cannot be blank " +
+          getContextInfo(name, source, tags));
+    }
+    if (source == null || source.isEmpty()) {
+      throw new IllegalArgumentException("event source or default source cannot be blank " +
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/Utils.java:304]: We should throw only if both source and default source are empty?

File: src/main/java/com/wavefront/sdk/common/clients/WavefrontClient.java [hunk 2]
```diff
@@ -388,6 +408,34 @@ public void sendDistribution(String name, List<Pair<Double, Integer>> centroids,
...
+    if (closed.get()) {
+      throw new IOException("attempt to send using closed sender");
+    }
+    String event;
+    URI uri = URI.create(this.clientId);
+    try {
+      if (uri.getScheme().equals(Constants.HTTP_PROXY_SCHEME)) {
+        event = eventToLineData(name, startMillis, endMillis, source, tags, annotations, defaultSource, false);
+      } else {
+        event = eventToLineData(name, startMillis, endMillis, source, tags, annotations, defaultSource, true);
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/clients/WavefrontClient.java:425]: let's add a comment here clarifying this path is direct ingestion.

File: src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java [hunk 3]
```diff
@@ -124,4 +180,18 @@ public static URL getReportingUrl(URI server, String format) throws MalformedURL
     URL url = new URL(server.getScheme(), server.getHost(), server.getPort(), originalPath + "?f=" + format);
     return url;
   }
+
+  @VisibleForTesting
+  public static URL getEventReportingUrl(URI server) throws MalformedURLException {
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java:185]: javadoc for the method

File: src/main/java/com/wavefront/sdk/direct/ingestion/WavefrontDirectIngestionClient.java [hunk 4]
```diff
@@ -382,6 +382,14 @@ private void sendSpanLogs(UUID traceId, UUID spanId, List<SpanLog> spanLogs, Str
     }
   }
 
+  public void sendEvent(String name, long startMillis, long endMillis, @Nullable String source,
+                        @Nullable Map<String, String> tags,
+                        @Nullable Map<String, String> annotations)
+      throws UnsupportedOperationException {
+    throw new UnsupportedOperationException("Sending event is not supported for " +
+        "deprecated WavefrontDirectIngestionClient.");
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/direct/ingestion/WavefrontDirectIngestionClient.java:390]: Also mention in the exception message hinting at which client to use.

File: src/main/java/com/wavefront/sdk/proxy/WavefrontProxyClient.java [hunk 5]
```diff
@@ -414,6 +414,15 @@ private void sendSpanLogsData(UUID traceId, UUID spanId, List<SpanLog> spanLogs,
     }
   }
 
+  @Override
+  public void sendEvent(String name, long startMillis, long endMillis, @Nullable String source,
+                        @Nullable Map<String, String> tags,
+                        @Nullable Map<String, String> annotations)
+      throws UnsupportedOperationException {
+    throw new UnsupportedOperationException("Sending event is not supported for " +
+        "deprecated WavefrontProxyClient.");
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/proxy/WavefrontProxyClient.java:423]: Ditto, mention the new client that supports this feature in the error message.

File: src/main/java/com/wavefront/sdk/common/clients/WavefrontClient.java [hunk 6]
```diff
@@ -388,6 +408,34 @@ public void sendDistribution(String name, List<Pair<Double, Integer>> centroids,
...
+                        @Nullable Map<String, String> tags,
+                        @Nullable Map<String, String> annotations)
+      throws IOException {
+    if (closed.get()) {
+      throw new IOException("attempt to send using closed sender");
+    }
+    String event;
+    URI uri = URI.create(this.clientId);
+    try {
+      if (uri.getScheme().equals(Constants.HTTP_PROXY_SCHEME)) {
```
Comment (vikramraman) [inline src/main/java/com/wavefront/sdk/common/clients/WavefrontClient.java:422]: Is the assumption here that if scheme is HTTP it is proxy. HTTPS is direct ingestion?

Comment (xshipeng) [reply src/main/java/com/wavefront/sdk/common/clients/WavefrontClient.java:422]: Yes. The path is formatted in https://github.com/wavefrontHQ/wavefront-sdk-java/blob/86ee8ba438101e104e4e3e15fb61d76e9a04cbc3/src/main/java/com/wavefront/sdk/common/clients/WavefrontClientFactory.java#L193. I will add the comment to clarify.

**Pull request merged**